model/labels/labels_internlabels.go

```diff
@@ -0,0 +1,778 @@
+			return lValue
+		} else if lName[0] > name[0] { // Stop looking if we've gone past.
+			break
+		}
+		_, i = decodeVarint(ls.data, i)
+	}
+	return ""
+}
+
+// Has returns true if the label with the given name is present.
+func (ls Labels) Has(name string) bool {
+	if name == "" { // Avoid crash in loop if someone asks for "".
+		return false // Prometheus does not store blank label names.
+	}
+	for i := 0; i < len(ls.data); {
+		var lName string
+		lName, i = decodeString(ls.syms, ls.data, i)
+		if lName == name {
+			return true
+		} else if lName[0] > name[0] { // Stop looking if we've gone past.
+			break
+		}
+		_, i = decodeVarint(ls.data, i)
+	}
+	return false
+}
+
+// HasDuplicateLabelNames returns whether ls has duplicate label names.
+// It assumes that the labelset is sorted.
+func (ls Labels) HasDuplicateLabelNames() (string, bool) {
+	prevNum := -1
+	for i := 0; i < len(ls.data); {
+		var lNum int
+		lNum, i = decodeVarint(ls.data, i)
+		_, i = decodeVarint(ls.data, i)
+		if lNum == prevNum {
+			return ls.syms.ToName(lNum), true
+		}
+		prevNum = lNum
+	}
+	return "", false
+}
+
+// WithoutEmpty returns the labelset without empty labels.
+// May return the same labelset.
+func (ls Labels) WithoutEmpty() Labels {
+	if ls.IsEmpty() {
+		return ls
+	}
+	// Idea: have a constant symbol for blank, then we don't have to look it up.
+	blank, ok := ls.syms.symbolTable.byName[""]
+	if !ok { // Symbol table has no entry for blank - none of the values can be blank.
+		return ls
+	}
+	for pos := 0; pos < len(ls.data); {
+		_, newPos := decodeVarint(ls.data, pos)
+		lValue, newPos := decodeVarint(ls.data, newPos)
+		if lValue != blank {
+			pos = newPos
+			continue
+		}
+		// Do not copy the slice until it's necessary.
+		// TODO: could optimise the case where all blanks are at the end.
+		// Note: we size the new buffer on the assumption there is exactly one blank value.
+		buf := make([]byte, pos, pos+(len(ls.data)-newPos))
+		copy(buf, ls.data[:pos]) // copy the initial non-blank labels
+		pos = newPos             // move past the first blank value
+		for pos < len(ls.data) {
+			var newPos int
+			_, newPos = decodeVarint(ls.data, pos)
+			lValue, newPos = decodeVarint(ls.data, newPos)
+			if lValue != blank {
+				buf = append(buf, ls.data[pos:newPos]...)
+			}
+			pos = newPos
+		}
+		return Labels{syms: ls.syms, data: yoloString(buf)}
+	}
+	return ls
+}
+
+// Equal returns whether the two label sets are equal.
+func Equal(a, b Labels) bool {
+	if a.syms == b.syms {
+		return a.data == b.data
+	}
+
+	la, lb := len(a.data), len(b.data)
+	ia, ib := 0, 0
+	for ia < la && ib < lb {
+		var aName, bName string
+		aName, ia = decodeString(a.syms, a.data, ia)
+		bName, ib = decodeString(b.syms, b.data, ib)
+		if aName != bName {
+			return false
+		}
+		var aValue, bValue string
+		aValue, ia = decodeString(a.syms, a.data, ia)
+		bValue, ib = decodeString(b.syms, b.data, ib)
+		if aValue != bValue {
+			return false
+		}
```

Comment:
Do we really need this bit of code? We know `a.data` has pairs of strings, we can use just one string decoding in the loop.